students/Nick_Haury/Lesson09/mailroom_oo/donor_models.py

```diff
@@ -0,0 +1,122 @@
+           "towards funding our common goal."
+           "\n\nEver Greatefully Yours,\n\n"
+           "X" + ("_" * 20) + "\n")
+    
+    def __init__(self, donor_name, donations = None):
+        if donations == None:
+            self._donations = []
+        else:
+            self._donations = donations
+        self._name = donor_name
+
+    @property
+    def name(self):
+        return self._name
+    
+    @property
+    def donations(self):
+        return self._donations
+
+    def email_text(self, donation_index):
+        try:
+            return self.email_template_string.format(
+                self.name, self.donations[donation_index])
+        except IndexError:
+            return None
+
+
+    def add_donation(self, donation_amount):
+        '''add a donation to the donor's list,
+        must be positive type int or float
+        '''
+        if isinstance(donation_amount, (int, float)):
+            if donation_amount > 0:
+                self.donations.append(donation_amount)
+            else:
+                raise ValueError("donation must be a positive value")
+        else:
+            raise TypeError('donation must be of type int or float')
+
+    def __str__(self):
+        return f'Donor named {self._name} with {len(self.donations)} donations.'
+
+    def __repr__(self):
+        return f'Donor("{self._name}", {self.donations})'
+
+    def __lt__(self, other):
+        '''sorts by sum of donations, then alphabetical by Donor name'''
+        return (sum(self.donations), self.name) < (sum(other.donations), other.name)
+
+
+class DonorCollection:
+
+    def __init__(self, donor_dict_input = {}):
+        if isinstance(donor_dict_input, dict):
+            self.donors = donor_dict_input
+        else:
+            raise TypeError("donor_dict_input must be a dictionary"
+                            "of form {'donor name':Donor()'}")
+
+    def add_donor(self, donor):
+        if isinstance(donor, Donor):
+            self.donors[donor.name] = donor
+        else:
+            raise TypeError("donor must be of Class Donor")
+
+    @property
+    def names(self):
+        return list(self.donors.keys())
+
+    @property
+    def lower_names(self):
+        return [name.lower() for name in self.names]
+
+    def report(self):
+        report_list = ["\nDonor Name" + " "*15 + "|  Total Given  "
+                         "| Num Gifts |   Average Gift\n"]
+        for donor in sorted(self.donors.values(), reverse=True):
+            report_list.append(f"{donor.name:26}"
+                               f"${sum(donor.donations):14,.2f}"
+                               f"{len(donor.donations):11}  "
+                               f"${st.mean(donor.donations):16,.2f}\n")
+        return report_list
+
+    def add_donation(self, donor_name, donation):
+        if donor_name.lower() in self.lower_names:
+            # check to see if different capitalization is being used
```

Comment:
you could store all names in lower case and capitalize on any display by using a property

Reply:
I was wondering about this... I was thinking I'd utilize the string.title() method, but it doesn't work for William Gates, III (results in William Gates, Iii).  Would have to create a specific property for that... I think I read about something called... regular expressions?  Maybe this is a good use case for that?